### PR TITLE
Make the vs environment take effect when generating the configuration…

### PR DESCRIPTION
**Pull request by deadash**
When meson generates the configuration, append_vcvars does not work. This pr submission is to make this parameter work.

The benefits are as follows:
- [x] A 32-bit program can be generated when vs is backend, which was not possible before
- [x] In the old version of meson, this allows meson to find vs

Signed-off-by: deadash <dead.ash@hotmail.com>

Changelog: (Feature): Describe here your pull request
Docs: https://github.com/conan-io/docs/pull/XXXX

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

### CODE REVIEW

**Comment by deadash** [comment]
The test is adjusted because the vc environment must be placed in the front, if it is placed in the back, there will be a problem of meson judgment error.

File: conans/test/functional/build_helpers/meson_test.py [hunk 1]
```diff
@@ -30,4 +30,5 @@ def run(self, cmd):
 
         client.save({"conanfile.py": conanfile_vcvars})
         client.run('create . pkg/1.0@ -e PATH="MyCustomPath"')
-        self.assertIn("pkg/1.0: PATH ENV VAR: MyCustomPath;", client.out)
+        self.assertIn("pkg/1.0: PATH ENV VAR: ", client.out)
```
Comment (memsharded) [inline conans/test/functional/build_helpers/meson_test.py:33]: How is this test different than before? Is this change necessary?

File: conans/client/build/meson.py [hunk 2]
```diff
@@ -164,7 +164,11 @@ def configure(self, args=None, defs=None, source_dir=None, build_dir=None,
             build_type
         ])
         command = 'meson "%s" "%s" %s' % (source_dir, self.build_dir, arg_list)
-        with environment_append({"PKG_CONFIG_PATH": pc_paths}):
+        envs = {"PKG_CONFIG_PATH": pc_paths}
+        if self._append_vcvars and str.startswith(self.backend, "vs"):
```
Comment (memsharded) [inline conans/client/build/meson.py]: why not ``self.backend.startswith("vs")``? This is the usual syntax we use in the codebase.

File: conans/client/build/meson.py [hunk 3]
```diff
@@ -164,7 +164,11 @@ def configure(self, args=None, defs=None, source_dir=None, build_dir=None,
...
+            vcvars_dict = tools.vcvars_dict(self._settings, output=self._conanfile.output)
```
Comment (memsharded) [inline conans/client/build/meson.py:169]: This change might be breaking. It introduces the usage of vcvars in an scenario that was not done before.

**Review by memsharded — COMMENTED** [review]
The whole helper is disabling, harcoded in code the usage of other backends:
```python
    def _run_ninja_targets(self, args=None, build_dir=None, targets=None):
        if self.backend != "ninja":
            raise ConanException("Build only supported with 'ninja' backend")
```
So this will not work. Adding tests trying to change the backed would fail.

Furthermore, this is the legacy integration. Changes and improvements should be targeted to the new ones: https://docs.conan.io/en/latest/reference/conanfile/tools.html

File: conans/client/build/meson.py (hunk 2 above)
Comment (deadash) [reply conans/client/build/meson.py]: Yes, there is no difference between the two, then do as you say.

I do this just for the convenience of automatic coding.

File: conans/test/functional/build_helpers/meson_test.py (hunk 1 above)
Comment (deadash) [reply conans/test/functional/build_helpers/meson_test.py:33]: Previously, the environment variable was inserted in the middle, and the vs environment variable must be inserted earlier, otherwise it will be destroyed by the other environment of the system.

The resulting effect is
```
pkg/1.0: PATH ENV VAR: [vs env];bla bla bla;MyCustomPath;
```
Instead of expected
```
pkg/1.0: PATH ENV VAR: MyCustomPath;[vs env];bla bla bla;
```

File: conans/client/build/meson.py (hunk 3 above)
Comment (deadash) [reply conans/client/build/meson.py:169]: This change is because when meson uses vs as the backend to generate vs solution, it is in the generation configuration stage. It happens that if vcvars is not attached, meson will find vcvars by itself, and then the 32-bit arch set will be invalid. In ninja, Environment variables are used in the generation phase, so ninja inserts vcvars when building.

E.g:
Set compiler.arch=x86
Then using vs as the backend, the 64-bit program is generated instead of the expected 32-bit program.
If you add this patch, VS will generate a 32-bit program.

**Comment by deadash** [comment]
@memsharded 

I tried a new compilation tool, which does not support vs as a backend. And there are some problems.

If you fully support vs as a backend, there are a lot of codes that need to be modified, such as whether to compile using meson completely, or only vs. using meson, and ninja uses ninja_target to compile.

The patch here does not adapt to this problem, but only fixes the problem when vs is used as a backend.

Just replace `meson.build()` with `meson._run_meson_command('compile', None, build_dir=self.build_folder)`

**Comment by deadash** [comment]
@memsharded Is there any reason why it cannot be merged? This pr will be very helpful for using meson instead of mesontoolchain.

**Comment by memsharded** [comment]
This PR has become outdated, I am closing it. 

Conan 2.0 has already removed the ``Meson`` legacy helper, and only the new ``from conan.tools.meson import Meson`` keeps there. Also Meson 1.0 is there, so probably some other things have changed too.

Thanks very much for your contribution!
